File: 0095-unique-binary-search-trees-ii/0095-unique-binary-search-trees-ii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    #def __init__(self):
    #    self.m = {}
    
    def solve(self, start, end):
        v=[]
        if start>end:
            v.append(None);
            return v
        if start==end:
            v.append(TreeNode(start,None,None))
            return v
        for i in range(start,end+1):
            l = self.solve(start,i-1)
            r = self.solve(i+1, end)
            for x in l:
                for y in r:
                    root = TreeNode(i)
                    root.left = x
                    root.right= y
                    v.append(root)
        return v
    
    def generateTrees(self, n: int) -> List[Optional[TreeNode]]:
        return self.solve(1,n)
        
        

File: 0095-unique-binary-search-trees-ii/0095-unique-binary-search-trees-ii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<TreeNode*> make_tree(int start, int end){
        vector<TreeNode*> v;
        if(start>end){
            v.push_back(nullptr);
            return v;
        }
        else if(start==end){
            v.push_back(new TreeNode(start));
            return v;
        }
        else{
            for(int i=start; i<=end; i++){
                vector<TreeNode*> l = make_tree(start, i-1);
                vector<TreeNode*> r = make_tree(i+1, end);
                
                for(auto le:l){
                    for(auto re:r){
                        TreeNode* root = new TreeNode(i);
                        root->left = le;
                        root->right = re;
                        v.push_back(root);
                    }
                }
            }
        }
        return v;
    }
    
    vector<TreeNode*> generateTrees(int n) {
        return make_tree(1,n);
    }
};